src/shutdown/umount.c

```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
                         continue;
 
                 /* Trying to umount */
-                if (umount_with_timeout(m, last_try) < 0)
+                r = umount_with_timeout(m, last_try);
+                if (r < 0)
                         n_failed++;
                 else
                         *changed = true;
+
+                /* If a mount is busy, we move it to not keep parent mount points busy.
+                 * If a mount point is not a leaf, moving it would invalidate our mount table.
+                 * More moving will occur in next iteration with a fresh mount table.
+                 *
+                 * umount_with_timeout will return EPROTO even on EBUSY due to forking.
+                 */
+                if (r == -EPROTO && m->leaf) {
+                        _cleanup_free_ char *dirname = NULL;
+
+                        r = path_extract_directory(m->path, &dirname);
+                        if (r < 0) {
+                                n_failed++;
+                                log_full_errno(last_try ? LOG_ERR : LOG_INFO, errno, "Cannot find directory for %s: %m", m->path);
+                                continue;
+                        }
+
+                        if (!path_equal_or_files_same(dirname, "/run/shutdown/mounts", 0)) {
+                                _cleanup_free_ char *newpath = NULL, *hexhash = NULL;
+                                uint8_t hash[SHA256_DIGEST_SIZE];
+
+                                sha256_direct(m->path, strlen(m->path), hash);
+
+                                hexhash = hexmem(hash, SHA256_DIGEST_SIZE);
+                                if (!hexhash)
+                                        return log_oom();
+
+                                asprintf(&newpath, "/run/shutdown/mounts/%s", hexhash);
+                                if (!newpath)
```

Comment:
this is likely going to trigger a static checker sooner or later, because you ignore the asprintf() return value here. hence do:

```c
if (asprintf(…) < 0)
        return log_oom();
```

Reply:
I have replaced it with `xsprintf` like proposed by @keszybz